### ISSUE TITLE
"import" can't be used in 'define' option

### ISSUE THREAD

**Issue by romainl** [vim/vim#1519]
## Background

JavaScript now has an official module mechanism where you can `export` a type from a module and `import` it into another:

    // filea.js
    const foo = 1;
    export { foo };

    // fileb.js
    import { foo as bar } from "filea";

## Problem

Using the word `import` in the `'define'` option has no effect.

## Reduced example

    import { foo as bar };
    export { foobar };
    console.log(foobar);

## What I tried

    :set define=^import
    :dlist /
    E388: Couldn't find definition

    :set define=^export
    :dlist /
    1    2    export { foobar };

    :set define=^\\(import\\\|export\\)
    :dlist /
    1    2    export { foobar };

**Comment by romainl** [comment]
After fiddling a bit with my config I came to the conclusion that `include` and `define` can't have matches on the same line.

Is this intended? How could we work around this situation and match the line below both as a `define` (we define a `bar` "thing")  and as an `include` (we include the file `filea.js`)?

    import { foo as bar } from "filea";

**Comment by brammool** [comment]
Romain Lafourcade wrote:

> After fiddling a bit with my config I came to the conclusion that `include` and `define` can't have matches on the same line.
> 
> Is this intended? How could we work around this situation and match the line below both as a `define` (we define a `bar` "thing")  and as an `include` (for `filea.js`)?
> 
>     import { foo as bar } from "filea";    

This appears to work:

   :set define=import\\s*{\\s*[a-z]*\\s\\+as\\s\\+

What do you set 'include' to then?  A reproducible example would be
useful.

-- 
"Space is big.  Really big.  You just won't believe how vastly hugely mind-
bogglingly big it is.  I mean, you may think it's a long way down the
road to the chemist, but that's just peanuts to space."
		-- Douglas Adams, "The Hitchhiker's Guide to the Galaxy"

 /// Bram Moolenaar -- Bram@Moolenaar.net -- http://www.Moolenaar.net   \\\
///        sponsor Vim, vote for features -- http://www.Vim.org/sponsor/ \\\
\\\  an exciting new programming language -- http://www.Zimbu.org        ///
 \\\            help me help AIDS victims -- http://ICCF-Holland.org    ///

**Comment by romainl** [comment]
Please ignore this. It was a PEBKAC issue: my `include` was sloppily written.

**Comment by romainl** [comment]
Hello Bram. I'm sorry to reopen this issue but I'm not sure anymore about its PEBKAC-ness.

Here is a simple reproducible example that shouldn't introduce any filetype-specific side-effects:

    $ mkdir test
    $ cd test
    $ echo 'import foo from "./export.txt";' > import.txt
    $ touch export.txt

With that setup, `'include'` works as expected on its own:

    $ vim -Nu NONE import.txt +set\ inc= +set\ def=
    :set include=^\\s*import\\s.*\\sfrom
    :checkpath!
        /path/to/test/export.txt
    :q

and so does `'define'`:

    $ vim -Nu NONE import.txt +set\ inc= +set\ def=
    :set define=^\\s*import
    :dlist foo
        import.txt
          1:    1 import foo from './export.txt';
    :q

But `'define'` doesn't seem to work when it matches on the same line as `'include'` (the order doesn't matter):

    $ vim -Nu NONE import.txt +set\ inc= +set\ def=
    :set include=^\\s*import\\s.*\\sfrom
    :checkpath!
        /path/to/test/export.txt
    :set define=^\\s*import
    :dlist foo
        E388: Couldn't find definition

If we add a new line with a random silly statement and adjust `'define'` accordingly, we can see that `'define'` works as expected:

    $ vim -Nu NONE import.txt +set\ inc= +set\ def=
    oratatouille bar = 1;<ESC>
    :set define=^\\s*ratatouille
    :dlist bar
        import.txt
          1:    2 ratatouille bar = 1;

Here are my observations:

* When both are set and match on the same line, `'include'` works but `'define'` doesn't.
* When both are set but match on different lines, both work as expected.
* When only `'include'` is set, it works as expected.
* When only `'define'` is set, it works as expected.

`'define'` doesn't seem to have any effect on `'include'` but `'include'` definitely seems to have some on `'define'`.

Are you able to reproduce this? If so, is this behavior intentional?

**Comment by bfrg** [comment]
I have encountered the same issue. Lines that match the `'include'` pattern are skipped in define-search.

Here's a minimal example:

```vim
set include=\\<require(
set define=^\\ze\\i\\+\ =
```

Now create the following two files:
```
# /tmp/mylib
foo = 10
```
and
```
# /tmp/main
x = require(mylib)
var = x.foo
      │
      └ put cursor here and press [d
```
In `/tmp/main` put the cursor on the second `x` and press `[d`, or alternatively, put the cursor on the second line and run `:dsearch x`. 

Result:
```
E388: Couldn't find definition
```
When pressing `[i`, Vim will print instead:
```
E387: Match is on current line
```

**Comment by brammool** [comment]
Yeah, the code does a depth-first search, thus entering the included file, and when coming back in the file with the "include" line continues on the next line.  Thus the "define" pattern isn't matched in the same line.  There is no trivial fix, this will require some work.
